### DISCUSSION TITLE
Error installing with Docker

### DISCUSSION THREAD

**Discussion by benjamin-cash** [dtcenter/METplus#2071] [Installation] [answered]
Hello,

I am trying to install Met using Docker, following the instructions here:
https://github.com/dtcenter/METbaseimage#how-to-use-dockerfiles

However, when I do that I get the following error:
docker build -t dtcenter/met-base:v2.0_debian10 -f Dockerfile.debian .
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /groups/BCASH/Dockerfile.debian: no such file or directory

Do I need to download a file to my local machine before running docker build? I'm pretty much completely unfamiliar with Docker so pardon me if that is obvious. :)

All the best,
Ben

**Comment by JohnHalleyGotway** [comment]
Hi Ben, @jprestop and @j-opatz gave me some context about your question. As I understand it, you'd like to try running MET via a Docker container rather than compiling it locally on your machine. And that's great! I'm glad that Docker is an option for you.

The good news is that using a pre-compiled Docker image is literally a one-liner. We have DockerHub configured to automatically create a new Docker image for each release we publish.

Here's MET's location on DockerHub: https://hub.docker.com/r/dtcenter/met

And here's the list of tags: https://hub.docker.com/r/dtcenter/met/tags

Pulling the latest bugfix version (11.0.1) from DockerHub to your machine is as simple as running:
```
docker pull dtcenter/met:11.0.1
```

There's no NEED for you to compile MET via Docker yourself. Instead, just pull a pre-built image.

**Comment by benjamin-cash** [comment]
Hi John,

When I run that command I get the following:

***@***.*** ~]$ docker pull dtcenter/met:11.0.1
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.24/images/create?fromImage=dtcenter%2Fmet&tag=11.0.1: dial unix /var/run/docker.sock: connect: permission denied

I can run just “docker pull” and get the message about options required, so it works up to that point at least. I’m not really familiar with docker so I don’t know how to debug from here.

All the best,
Ben

**Reply by JohnHalleyGotway** [reply]
Ben, yeah, you'll need some level of knowledge of Docker to make good use of the METplus-related Docker images.

When getting started with Docker, it's a good idea to run:
```
docker run hello-world
```

And that should print the following:
```
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
2db29710123e: Pull complete 
Digest: sha256:6e8b6f026e0b9c419ea0fd02d3905dd0952ad1feea67543f525c73a0a790fefb
Status: Downloaded newer image for hello-world:latest

Hello from Docker!
This message shows that your installation appears to be working correctly.
```

Doing some Googling about your error led me to this Docker documentation:
https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user

Hopefully that direction is helpful.

If you'd like to learn more about running Docker containers for NWP, you may find this online tutorial helpful:
https://dtcenter.org/nwp-containers-online-tutorial

We developed this over the last several years. The project is no longer active/funded, but the command *should* still work.

A second option is this video I created to demonstrate running METviewer via a Docker container:
https://metplus-training.readthedocs.io/en/latest/modules/METviewer/docker.html

**Comment by benjamin-cash** [comment]
I should also point out that these instructions are different than the ones given here https://github.com/dtcenter/METbaseimage#how-to-use-dockerfiles

**Reply by JohnHalleyGotway** [reply]
Ben, I suppose I'm not surprised that there's some confusion here. We use Docker in multiple ways to do a variety of things in METplus.

1. To provide software releases on DockerHub for [dtcenter/MET](https://hub.docker.com/repository/docker/dtcenter/met/general), [dtcenter/METplus](https://hub.docker.com/repository/docker/dtcenter/metplus/general), and [dtcenter/METveiwer](https://hub.docker.com/repository/docker/dtcenter/metviewer/general). **This is what you want and care about.**
2. To facilitate our continuous integration testing via GitHub actions using DockerHub [dtcenter/MET-dev](https://hub.docker.com/repository/docker/dtcenter/met-dev/general) and [dtcenter/METplus-dev](https://hub.docker.com/repository/docker/dtcenter/metplus-dev/general).
3. To host MET's input data ([dtcenter/MET-data](https://hub.docker.com/repository/docker/dtcenter/met-data/general)) and output data ([dtcenter/MET-data](https://hub.docker.com/repository/docker/dtcenter/met-data-output/general)) for CI testing.
4. To host METplus's input data ([dtcenter/METplus-data](https://hub.docker.com/repository/docker/dtcenter/metplus-data/general)), output data ([dtcenter/METplus-data-dev](https://hub.docker.com/repository/docker/dtcenter/metplus-data-dev/general)), and python environments ([dtcenter/METplus-envs](https://hub.docker.com/repository/docker/dtcenter/metplus-envs/general)) for CI testing.

Docker is very powerful and useful but there's plenty of room for confusion. You've been asking about METbaseimage (in [GitHub](https://github.com/dtcenter/METbaseimage) and on [DockerHub](https://hub.docker.com/repository/docker/dtcenter/met-base/general)). MET has many dependencies (i.e. NetCDF, GSL, PrepBUFR, GRIB2, and so on). Rather than recompiling all of those each time we want to recompile MET or redistribute METviewer, we serve them up via pre-compiled base image. That's what this is for. So instead of starting from scratch, we start with an environment that's all ready to compile MET. That makes things run faster.

From a user perspective, all you care about is running released versions of the software. And (once your Docker environment is up and running), that's just an easy one-liner.

I'm glad you're on the "focus group" with @j-opatz and can provide some feedback about clarifying/simplifying instructions for getting up and running with MET/METviewer/METplus via Docker.

**Reply by benjamin-cash** [reply]
Hi @JohnHalleyGotway - The link that I pointed to is the link that was sent out to me as a member of the focus group specifically as an alternative way to install metplus. If that is not the correct procedure for what I need to do as someone who just wants to use metplus then that email should be updated to point to the correct commands.

**Reply by JohnHalleyGotway** [reply]
@j-opatz, just calling your attention to this feedback from @benjamin-cash. I presume you're the person who sent out info to the focus group.

**Reply by j-opatz** [reply]
I was, and I'll send out a clarification email now with the proper webpage.

**Reply by benjamin-cash** [reply]
The good news is that I figured out what was wrong with my first attempt to install met - our cluster doesn't actually use docker, and it's a bit of a mystery how I got to docker at all right now. It uses singularity, and after a little poking around I was able to get it to work with the following:
singularity build met-11.0.1 docker://dtcenter/met:11.0.1

as an FYI, met:latest does not point to 11.0.1, when I built that I got:
Singularity> point_stat

*** Model Evaluation Tools (METV10.1.2) ***

**Reply by JohnHalleyGotway** [reply]
Great, and thanks for pointing out what "latest" contains. I'm checking in with @georgemccabe about differences I see in the DockerHub auto-build rules for MET vs METplus. While METplus DOES still have a rule to build a "latest" tag, MET does NOT. We need to discuss why we removed that. Once we rectify those differences, we'll either remove the existing "latest" tags from both... or make sure they're present for both.

**Comment by benjamin-cash** [comment, marked as answer]
Just pulling out what ultimately ended up being the answer for my system using singularity:
singularity build met-11.0.1 docker://dtcenter/met:11.0.1
